### PR TITLE
WIP: Trimmed statistics functions have inconsistent API.

### LINKED ISSUE #2914: Trimmed statistics functions have inconsistent API
These functions are:

- tmin
- tmax
- tmean
- tstd
- tvar
- tsem

They should all be made consistent, probably all need keywords ``axis`` and ``ddof``. ``tsem`` should be made consistent with ``sem``. The relation between ``tmean`` and ``trim_mean`` should be clarified.

Those functions are now all tested to some extent after gh-2913. What's left to do there is to test their behavior with masked arrays. Examples should also still be added.

### PR DESCRIPTION
**Pull request by aeklant**
closes gh-2914

### CODE REVIEW

**Comment by argriffing** [comment]
The TravisCI error seems to be a timeout related to coverage after passing the tests.

**Comment by aeklant** [comment]
@argriffing Does that mean the error can be disregarded?

**Comment by argriffing** [comment]
> Does that mean the error can be disregarded?

Yes I think so.  This is related to https://github.com/scipy/scipy/issues/3913 - SciPy takes so long to build and test that it sporadically hits the TravisCI 50 minute time limit.  The error can be disregarded, or TravisCI could be restarted, or the problem could be patched by breaking up the tests or switching CI providers or probably by giving money to TravisCI.

**Comment by rgommers** [comment]
These changes look good to me, signatures are correct. Too bad about ``ddof=1`` (numpy uses 0) but that's necessary for backwards compat.

**Comment by aeklant** [comment]
I included `trim*`because it seemed to make more sense than having a bunch of small related PR's instead of just one to keep track. I think the only things pending are masked array related, but I will get to that later on.

**Comment by rgommers** [comment]
Makes sense to add ``trim*`` functions here. I've put some notes on those functions in gh-620.

**Comment by aeklant** [comment]
I addressed the comments. Should I run the backwards incompatible change of `trim1` and `trimboth` by the mailing list?

**Comment by rgommers** [comment]
@aeklant yes, would be good to discuss those on the mailing list.

File: doc/release/0.17.0-notes.rst [hunk 1]
```diff
@@ -29,6 +29,9 @@ Deprecated features
 Backwards incompatible changes
 ==============================
 
+The functions ``trim1`` and ``trimboth`` now sort elements before trimming.
+This is backwards incompatible but it made no sense to trim unsorted input.
```
Comment (juliantaylor) [inline doc/release/0.17.0-notes.rst]: full sorting is not necessary, a partial sort is enough, see `np.partition` available since 1.8. As a fallback you can use a full sort, but I wouldn't document the result to be fully sorted.

Comment (aeklant) [inline doc/release/0.17.0-notes.rst]: @juliantaylor yes, you can see `np.partition` being used as you mention in `trim_mean` because the ordering really doesn't matter either way. 

For `trim1` and `trimboth` there is a tradeoff:
**`np.sort`:** 
*  For small sized input (e.g. 1000 elements) the difference in performance between `np.sort` and `np.partition` seems negligible.
* Makes more sense in my opinion because this is not an in-place change (i.e. the result doesn't change the input array, but produce a new one) so it seems better to have the new result already sorted.

**`np.partition`:**
* For larger sized input (e.g. ~100,000 elements) `np.sort` time was around 2.5x that of `np.partition`

The tie braker was that having sorted output made more sense to me, but this is obviously an opinion. More feedback is appreciated.

Comment (juliantaylor) [inline doc/release/0.17.0-notes.rst]: the larger the array the faster partition is than sort as its a linear scaling instead of a linarithmic one, for even for 1000 elements I still get a factor 2 better performance.

I don't expect of a function named `trim` to sort the data, its not called `trim_and_sort`

Also note that `trim_mean` now sorts/partitions twice as it calls `trimboth`

Comment (sturlamolden) [inline doc/release/0.17.0-notes.rst]: ```np.partition``` will scale better, but as long as we support NumPy 1.6 and 1.7 we need to have a fallback method for those.

**Comment by aeklant** [comment]
Since everyone who showed some interest seems to agree to the partial sorting I made that change. I also corrected the docstrings to reflect that change.

I had to make sure I sorted the output in the tests to avoid tests failing depending on whether they are run with numpy >= 1.8.0 or not. I could have also checked for the numpy version in each of the tests but this seemed easier to do and I didn't think there was a need to get technical since they're tests. Let me know what you think about that (e.g. Is the impact in time for TravisCI worth checking for the version instead of sorting?).

**Comment by juliantaylor** [comment]
I wouldn't document the exact ordering of the result, as that is an implementational detail.
I'd just say the ordering of the trimmed content is undefined.
The tests should also not need to care about running in with partition or without, they only need to test that the results have been correctly trimmed, nothing else.

also trim_mean still sorts twice.

**Comment by sturlamolden** [comment]
Yes, ```trim_mean``` should not call ```trimboth``` to avoid sorting the array twice.

File: doc/release/0.17.0-notes.rst [hunk 2]
```diff
@@ -29,6 +29,11 @@ Deprecated features
 Backwards incompatible changes
 ==============================
 
+The functions ``trim1`` and ``trimboth`` now partially (for numpy >= 1.8.0) or
```
Comment (rgommers) [inline doc/release/0.17.0-notes.rst]: Can you make this ``stats.trim1`` and ``stats.trimboth``? Otherwise they may be hard to find.

File: doc/release/0.17.0-notes.rst [hunk 3]
```diff
@@ -29,6 +29,11 @@ Deprecated features
...
+fully (for numpy < 1.8.0) sort elements before trimming. This to make sure 
+the trimmed elements are the lowest or highest depending on the case. 
+Slicing without at least partial sorting was random hacking.
```
Comment (rgommers) [inline doc/release/0.17.0-notes.rst]: I'd rephrase that a bit: "was random hacking" --> "was previously done, but didn't make sense for unsorted input"

**Comment by sturlamolden** [comment]
Good, this only sorts once.

**Comment by rgommers** [comment]
@aeklant I see one last thing that would be good to address: ``tmax`` and ``tmin`` return a masked array, while the other functions all return float or ndarray. This is inconsistent, and there is no reason for masked arrays to be returned anywhere in ``stats.py`` (it's an implementation detail leaking out). Could you change that and update the docstring examples to match?

Simple way would be to change last line to:

    res = ma.maximum.reduce(am, axis).data
    if res.ndim == 0:
        return res[()]
    
    return res

**Comment by aeklant** [comment]
Is there something else to review? Do you guys think it looks good for merging now?

**Comment by rgommers** [comment]
Looks good to me now, merging. Thanks @aeklant, @jaimefrio

**Comment by aeklant** [comment]
Thanks @rgommers. Can you close `trim*` issues: gh-618, gh-619, gh620 as they were handled in this PR too.

**Comment by rgommers** [comment]
Sure, done.
